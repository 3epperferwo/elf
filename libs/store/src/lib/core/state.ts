type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export type State<State, Config> = { state: State; config: Config };
export type EmptyConfig = Record<string, any>;

export function createState<State extends any[]>(
  ...state: State
): { state: Merge<State, 'state'>; config: Merge<State, 'config'> } {
  return state.reduce(
    (acc, current) => {
      acc.config = {
        ...acc.config,
        ...current.config,
      };

      acc.state = {
        ...acc.state,
        ...current.state,
      };

      return acc;
    },
    { config: {}, state: {} }
  );
}
