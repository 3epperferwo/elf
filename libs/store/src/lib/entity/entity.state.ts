import { Store } from '@eleanor/store';

export const defaultEntitiesKey = '$entities' as const;
export const defaultIdsKey = '$ids' as const;

export class EntitiesRef<EntitiesKey extends string = any, IdsKey extends string = any> {
  entitiesKey: EntitiesKey;
  idsKey: IdsKey;

  constructor(private config: { entitiesKey: EntitiesKey, idsKey: IdsKey }) {
    this.entitiesKey = config.entitiesKey ?? defaultEntitiesKey;
    this.idsKey = config.idsKey ?? defaultIdsKey;
  }
}

export const defaultEntitiesRef = new EntitiesRef({ entitiesKey: defaultEntitiesKey, idsKey: defaultIdsKey });

export const entitiesUIRef = new EntitiesRef({
  entitiesKey: '$UIEntities',
  idsKey: '$UIIds'
});

export function withEntitiesFactory<S extends EntitiesRecord,
  Ref extends EntitiesRef,
  >(ref: Ref) {
  return {
    [ref.entitiesKey]: {},
    [ref.idsKey]: []
  } as unknown as {
    [P in Ref['entitiesKey'] | Ref['idsKey']]: S[P]
  };
}

export function withEntities<EntityType, IdType extends PropertyKey>(config: Partial<Config> = {}) {
  return {
    state: withEntitiesFactory<EntityState<EntityType, IdType>, typeof defaultEntitiesRef>(defaultEntitiesRef),
    config: {
      idKey: config?.idKey ?? 'id'
    }
  };
}

export function withUIEntities<EntityType, IdType extends PropertyKey>(config: Partial<Config> = {}) {
  return {
    state: withEntitiesFactory<UIEntityState<EntityType, IdType>, typeof entitiesUIRef>(entitiesUIRef),
    config: {
      idKey: config?.idKey ?? 'id'
    }
  };
}

export function getIdKey<T>(store: Store): T {
  return store.getConfig<Config>().idKey as unknown as T;
}

export interface EntityState<EntityType = any, IdType extends PropertyKey = any> {
  $entities: Record<IdType, EntityType>;
  $ids: Array<IdType>;
}

export interface UIEntityState<EntityType = any,
  IdType extends PropertyKey = any> {
  $UIEntities: Record<IdType, EntityType>;
  $UIIds: Array<IdType>;
}

interface Config {
  idKey: string;
}

export type getEntityType<S extends EntitiesRecord, Ref extends EntitiesRef> = S[Ref['entitiesKey']][0];
export type getIdType<S extends EntitiesRecord, Ref extends EntitiesRef> = S[Ref['idsKey']][0];
export type ItemPredicate<Item> = (item: Item, index?: number) => boolean;
export type EntitiesRecord = Record<any, any>;
export type DefaultEntitiesRef = typeof defaultEntitiesRef;

export interface BaseEntityOptions<Ref extends EntitiesRef> {
  ref?: Ref;
}
