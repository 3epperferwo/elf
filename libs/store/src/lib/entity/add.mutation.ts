import { BaseEntityOptions, DefaultEntitiesRef, defaultEntitiesRef, EntitiesRecord, EntitiesRef, getEntityType, getIdKey, getIdType } from './entity.state';
import { coerceArray } from '../core/utils';
import { OrArray } from '../core/types';
import { Reducer, Store } from '../core/store';

/**
 *
 * Add entities to the store
 *
 * store.reduce(addEntity(entity))
 *
 * store.reduce(addEntity([entity, entity]))
 *
 * store.reduce(addEntity([entity, entity]), { overwrite: true })
 *
 * store.reduce(addEntity([entity, entity]), { prepend: true })
 */
export function addEntity<S extends EntitiesRecord, Ref extends EntitiesRef = DefaultEntitiesRef>(
  entities: OrArray<getEntityType<S, Ref>>,
  options: {
    prepend?: boolean;
    overwrite?: boolean;
  } & BaseEntityOptions<Ref> = {}
): Reducer<S> {
  return function reducer(state: S, store: Store) {
    const { overwrite = false, prepend = false, ref = defaultEntitiesRef } = options;

    const { entitiesKey, idsKey } = ref!;

    const toObject = {} as Record<getIdType<S, Ref>, getEntityType<S, Ref>>;
    const ids = [] as getIdType<S, Ref>;
    const idKey = getIdKey<getIdType<S, Ref>>(store);

    for(const entity of coerceArray(entities)) {
      const id = entity[idKey];
      if(!overwrite && id in state[entitiesKey]) continue;

      ids.push(id);
      toObject[id] = entity;
    }

    if(!ids.length) return state;

    return {
      ...state,
      [entitiesKey]: overwrite ? toObject : { ...state[entitiesKey], ...toObject },
      [idsKey]: overwrite ? ids : prepend ? [...ids, ...state[idsKey]] : [...state[idsKey], ...ids]
    };
  };
}
