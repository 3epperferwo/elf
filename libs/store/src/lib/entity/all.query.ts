import { EntityState, getEntityType } from './entity.state';
import { MonoTypeOperatorFunction, OperatorFunction, pipe } from 'rxjs';
import { distinctUntilChanged, map } from 'rxjs/operators';

export function untilEntitiesChanges<T extends EntityState>(): MonoTypeOperatorFunction<T> {
  return distinctUntilChanged(
    (prev, current) => prev.$entities === current.$entities
  );
}

export function selectAll<S extends EntityState>(): OperatorFunction<S, getEntityType<S>[]> {
  return pipe(
    untilEntitiesChanges(),
    map(state => state.$ids.map((id) => state.$entities[id]))
  );
}
