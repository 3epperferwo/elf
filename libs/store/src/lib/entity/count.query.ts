import { OperatorFunction, pipe } from 'rxjs';
import { distinctUntilChanged, map } from 'rxjs/operators';
import { EntityState, getEntityType, ItemPredicate } from './entity.state';
import { untilEntitiesChanges } from './all.query';
import { select } from '../core/queries';

/**
 * Select the store's entity collection length
 *
 * @example
 *
 * store.pipe(selectCount())
 *
 */
export function selectCount<T extends EntityState>(): OperatorFunction<T, number> {
  return select(state => state.$ids.length);
}

/**
 * Select the store's entity collection length
 *
 * @example
 *
 * store.pipe(selectCount(entity => entity.completed))
 *
 */
export function selectCountByPredicate<T extends EntityState>(predicate: ItemPredicate<getEntityType<T>>): OperatorFunction<T, number> {
  return pipe(
    untilEntitiesChanges(),
    map(state => state.$ids.filter((id) => predicate(state.$entities[id])).length),
    distinctUntilChanged()
  );
}
