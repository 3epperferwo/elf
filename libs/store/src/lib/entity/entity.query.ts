import { OperatorFunction } from 'rxjs';
import { select } from '../core/queries';
import { isString, isUndefined } from '../core/utils';
import { BaseEntityOptions, defaultEntitiesRef, DefaultEntitiesRef, EntitiesRecord, EntitiesRef, getEntityType, getIdType } from './entity.state';

interface Options extends BaseEntityOptions<any> {
  pluck?: string | ((entity: unknown) => any);
}

/**
 * Select entity from the store
 *
 * @example
 *
 * store.pipe(selectEntity(id, { pluck: 'title' })
 *
 */
export function selectEntity<S extends EntitiesRecord, K extends keyof getEntityType<S, Ref>, Ref extends EntitiesRef = DefaultEntitiesRef>(
  id: getIdType<S, Ref>,
  options: { pluck: K } & BaseEntityOptions<Ref>
): OperatorFunction<S, getEntityType<S, Ref>[K]>;

/**
 * Select entity from the store
 *
 * @example
 *
 * store.pipe(selectEntity(id, { pluck: e => e.title })
 *
 */
export function selectEntity<S extends EntitiesRecord, R, Ref extends EntitiesRef = DefaultEntitiesRef>(
  id: getIdType<S, Ref>,
  options: { pluck: (entity: getEntityType<S, Ref>) => R } & BaseEntityOptions<Ref>
): OperatorFunction<S, R>;

/**
 * Select entity from the store
 *
 * @example
 *
 * store.pipe(selectEntity(id)
 *
 */
export function selectEntity<S extends EntitiesRecord, Ref extends EntitiesRef = DefaultEntitiesRef>(
  id: getIdType<S, Ref>,
  options?: BaseEntityOptions<Ref>
): OperatorFunction<S, getEntityType<S, Ref>>;

export function selectEntity<S extends EntitiesRecord, R>(id: any, options: Options = {}) {
  const { ref: { entitiesKey, pluck } = defaultEntitiesRef } = options;

  return select<S, R>(state => getEntity(state[entitiesKey], id, pluck));
}

export function getEntity(
  entities: EntitiesRecord,
  id: string | number,
  pluck: Options['pluck']
) {
  const entity = entities[id];

  if(isUndefined(entity)) {
    return undefined;
  }

  if(!pluck) {
    return entity;
  }

  if(isString(pluck)) {
    return entity[pluck];
  }

  return pluck(entity);
}
