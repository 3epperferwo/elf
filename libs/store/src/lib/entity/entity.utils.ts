import { EntitiesRecord, EntitiesRef, getEntityType, getIdType } from './entity.state';
import { coerceArray } from '../core/utils';

export function buildEntities<S extends EntitiesRecord, Ref extends EntitiesRef>(ref: Ref, entities: getEntityType<S, Ref>[], idKey: string):
  { ids: getIdType<S, Ref>[], asObject: getEntityType<S, Ref> } {
  const asObject = {} as Record<getIdType<S, Ref>, getEntityType<S, Ref>>;
  const ids = [] as getIdType<S, Ref>;

  for(const entity of coerceArray(entities)) {
    const id: getIdType<S, Ref> = entity[idKey];

    ids.push(id);
    asObject[id] = entity;
  }

  return {
    ids,
   asObject
  } as unknown as { ids: getIdType<S, Ref>[], asObject: getEntityType<S, Ref> };
}
