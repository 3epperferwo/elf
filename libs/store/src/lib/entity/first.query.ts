import { select } from '../core/queries';
import { EntityState, getEntityType } from './entity.state';

export function selectFirst<S extends EntityState>() {
  return select<S, getEntityType<S> | undefined>(state => state.$entities[state.$ids[0]]);
}
