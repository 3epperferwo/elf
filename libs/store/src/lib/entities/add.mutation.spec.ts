import { createEntitiesStore, createTodo, createUIEntityStore, createUITodo, toMatchSnapshot } from '../mocks/stores.mock';
import { addEntities } from './add.mutation';
import { entitiesUIRef } from './entity.state';

describe('add', () => {
  let store: ReturnType<typeof createEntitiesStore>;

  beforeEach(() => {
    store = createEntitiesStore();
  });

  it('should add entity', () => {
    store.reduce(addEntities(createTodo(1)));
    toMatchSnapshot(expect, store);
  });

  it('should add multiple entities', () => {
    store.reduce(addEntities([createTodo(1), createTodo(2)]));
    toMatchSnapshot(expect, store);
  });

  it('should prepend entities', () => {
    store.reduce(addEntities([createTodo(1), createTodo(2)]));
    store.reduce(addEntities([createTodo(3), createTodo(4)], { prepend: true }));
    toMatchSnapshot(expect, store);
  });

  it('should work with ref', () => {
    const store = createUIEntityStore();
    store.reduce(addEntities([createUITodo(1), createUITodo(2)], { ref: entitiesUIRef }));
    toMatchSnapshot(expect, store);
  });

});
