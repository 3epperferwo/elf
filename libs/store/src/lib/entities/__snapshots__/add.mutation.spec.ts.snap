// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add should add entity: add one 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "ids": Array [
    1,
  ],
}
`;

exports[`add should add multiple entities: add two 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
  },
  "ids": Array [
    1,
    2,
  ],
}
`;

exports[`add should prepend entities: prepend 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
    "2": Object {
      "completed": false,
      "id": 2,
      "title": "todo 2",
    },
    "3": Object {
      "completed": false,
      "id": 3,
      "title": "todo 3",
    },
    "4": Object {
      "completed": false,
      "id": 4,
      "title": "todo 4",
    },
  },
  "ids": Array [
    3,
    4,
    1,
    2,
  ],
}
`;

exports[`add should work with ref: ref 1`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
    "2": Object {
      "id": 2,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
    2,
  ],
}
`;
